core/blockchain.go

```diff
@@ -201,6 +201,10 @@ func NewBlockChain(db ethdb.Database, cacheConfig *CacheConfig, chainConfig *par
 	if bc.genesisBlock == nil {
 		return nil, ErrNoGenesis
 	}
+	var nilBlock *types.Block
+	bc.currentBlock.Store(nilBlock)
```

Comment:
Would it be better to put `bc.genesisBlock` here?

Reply:
I am not convinced that would be correct. That would change the semantics, and make it appear as if the current block is genesis -- which has a different meaning than `nil`. The latter indicates it is simply not set (yet), the former indicates that all is in order, we should start importing from genesis.

Reply:
This is actually what is already done in [`HeaderChain`](https://github.com/ethereum/go-ethereum/blob/master/core/headerchain.go#L100) for `HeaderChain.currentHeader`, so I don't think this changes semantics.

My understanding is that `BlockChain.loadLastState()` will overwrite this value because it reads directly from the database as well. IMHO there should be an invariant where `BlockChain.currentBlock` is always set.

Also note that the code in #19286 is attempting to set the head to genesis block anyways.

Reply:
It is also worth wondering how it got into this state. The chain configuration and genesis block were already present or `SetupGenesisBlockWithOverride` would have handled this, so maybe the "pull the plug" was when the head block hash key in the database got corrupted in #19286?